File: backend/app/services/workflow_service.py
"""Workflow state-machine service.

Centralizes valid status transitions, transition validation, and workflow
event creation for tasks.
"""
from __future__ import annotations

from typing import Dict, List, Optional, Set

from fastapi import HTTPException, status as http_status
from sqlalchemy.orm import Session

from app.models.task import Task, TaskStatus
from app.models.user import User
from app.models.workflow_event import WorkflowEvent
from app.repositories.task_repository import TaskRepository


# Allowed task status transitions. The single source of truth.
VALID_TRANSITIONS: Dict[TaskStatus, Set[TaskStatus]] = {
    TaskStatus.PENDING: {TaskStatus.IN_PROGRESS, TaskStatus.SUBMITTED},
    TaskStatus.IN_PROGRESS: {TaskStatus.SUBMITTED},
    TaskStatus.SUBMITTED: {TaskStatus.APPROVED, TaskStatus.REJECTED},
    TaskStatus.APPROVED: {TaskStatus.COMPLETED},
    TaskStatus.REJECTED: {TaskStatus.IN_PROGRESS, TaskStatus.SUBMITTED},
    TaskStatus.COMPLETED: set(),
}


class WorkflowService:
    """Validates status transitions and records workflow events."""

    def __init__(self, db: Session) -> None:
        self.db = db
        self.repo = TaskRepository(db)

    @staticmethod
    def is_valid_transition(from_status: TaskStatus, to_status: TaskStatus) -> bool:
        """Return True if the given transition is allowed."""
        return to_status in VALID_TRANSITIONS.get(from_status, set())

    def assert_valid_transition(self, from_status: TaskStatus, to_status: TaskStatus) -> None:
        """Raise an HTTP 400 if the transition is not allowed."""
        if not self.is_valid_transition(from_status, to_status):
            raise HTTPException(
                status_code=http_status.HTTP_400_BAD_REQUEST,
                detail=(
                    f"Invalid status transition: {from_status.value} -> {to_status.value}"
                ),
            )

    def transition(
        self,
        task: Task,
        to_status: TaskStatus,
        changed_by: User,
        note: Optional[str] = None,
    ) -> WorkflowEvent:
        """Validate the transition, mutate the task, and emit a workflow event."""
        from_status = task.status
        self.assert_valid_transition(from_status, to_status)
        task.status = to_status
        event = WorkflowEvent(
            task_id=task.id,
            from_status=from_status,
            to_status=to_status,
            changed_by_user_id=changed_by.id,
            note=note,
        )
        self.db.add(event)
        self.db.flush()
        return event

    def record_initial_event(
        self,
        task: Task,
        changed_by: User,
        note: Optional[str] = None,
    ) -> WorkflowEvent:
        """Record an event for task creation (no previous status)."""
        event = WorkflowEvent(
            task_id=task.id,
            from_status=None,
            to_status=task.status,
            changed_by_user_id=changed_by.id,
            note=note or "Task created",
        )
        self.db.add(event)
        self.db.flush()
        return event

    def get_task_history(self, task_id: int) -> List[WorkflowEvent]:
        return self.repo.list_workflow_events(task_id)

File: backend/src/main/java/com/example/bwms/service/WorkflowService.java
// Translated from: backend/app/services/workflow_service.py
package com.example.bwms.service;

import com.example.bwms.model.*;
import com.example.bwms.repository.WorkflowEventRepository;
import org.springframework.http.HttpStatus;
import org.springframework.lang.Nullable;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.util.EnumMap;
import java.util.EnumSet;
import java.util.List;
import java.util.Map;

@Service
public class WorkflowService {

    // Python: VALID_TRANSITIONS dict — single source of truth for allowed state changes
    public static final Map<TaskStatus, EnumSet<TaskStatus>> VALID_TRANSITIONS;
    static {
        VALID_TRANSITIONS = new EnumMap<>(TaskStatus.class);
        VALID_TRANSITIONS.put(TaskStatus.PENDING,    EnumSet.of(TaskStatus.IN_PROGRESS, TaskStatus.SUBMITTED));
        VALID_TRANSITIONS.put(TaskStatus.IN_PROGRESS, EnumSet.of(TaskStatus.SUBMITTED));
        VALID_TRANSITIONS.put(TaskStatus.SUBMITTED,  EnumSet.of(TaskStatus.APPROVED, TaskStatus.REJECTED));
        VALID_TRANSITIONS.put(TaskStatus.APPROVED,   EnumSet.of(TaskStatus.COMPLETED));
        VALID_TRANSITIONS.put(TaskStatus.REJECTED,   EnumSet.of(TaskStatus.IN_PROGRESS, TaskStatus.SUBMITTED));
        VALID_TRANSITIONS.put(TaskStatus.COMPLETED,  EnumSet.noneOf(TaskStatus.class));
    }

    private final WorkflowEventRepository workflowEventRepository;

    public WorkflowService(WorkflowEventRepository workflowEventRepository) {
        this.workflowEventRepository = workflowEventRepository;
    }

    // Python: WorkflowService.is_valid_transition (static)
    public static boolean isValidTransition(TaskStatus from, TaskStatus to) {
        EnumSet<TaskStatus> allowed = VALID_TRANSITIONS.get(from);
        return allowed != null && allowed.contains(to);
    }

    // Python: WorkflowService.assert_valid_transition → HTTP 400
    public static void assertValidTransition(TaskStatus from, TaskStatus to) {
        if (!isValidTransition(from, to)) {
            throw new ResponseStatusException(
                    HttpStatus.BAD_REQUEST,
                    "Invalid status transition: " + from.getValue() + " -> " + to.getValue()
            );
        }
    }

    // Python: WorkflowService.transition — mutates task, emits WorkflowEvent
    // @Transactional boundary owned by the calling service method
    public WorkflowEvent transition(Task task, TaskStatus toStatus,
                                    User changedBy, @Nullable String note) {
        TaskStatus fromStatus = task.getStatus();
        assertValidTransition(fromStatus, toStatus);
        task.setStatus(toStatus);
        WorkflowEvent event = new WorkflowEvent(
                task.getId(), fromStatus, toStatus, changedBy.getId(), note);
        return workflowEventRepository.save(event);
    }

    // Python: WorkflowService.record_initial_event — fromStatus is null for task creation
    public WorkflowEvent recordInitialEvent(Task task, User changedBy, @Nullable String note) {
        WorkflowEvent event = new WorkflowEvent(
                task.getId(), null, task.getStatus(),
                changedBy.getId(), note != null ? note : "Task created");
        return workflowEventRepository.save(event);
    }

    public List<WorkflowEvent> getTaskHistory(Long taskId) {
        return workflowEventRepository.findByTaskIdOrderByCreatedAtAsc(taskId);
    }
}
